homeassistant/components/recorder/__init__.py

```diff
@@ -556,11 +556,27 @@ def _async_setup_periodic_tasks(self):
         async_track_time_change(
             self.hass, self.async_nightly_tasks, hour=4, minute=12, second=0
         )
+
         # Compile hourly statistics every hour at *:12
         async_track_time_change(
             self.hass, self.async_hourly_statistics, minute=12, second=0
         )
 
+        # Add tasks for missing statistics runs
+        now = dt_util.utcnow()
+        start = now - timedelta(days=self.keep_days)
+        start = start.replace(minute=0, second=0, microsecond=0)
+        if not self.get_session:
+            # Home Assistant is shutting down
+            return
+        with session_scope(session=self.get_session()) as session:
+            while start < now:
+                if not session.query(StatisticsRuns).filter_by(start=start):
```

Comment:
Better to query all statistic where `run.start > oldest_start` at once and compare in memory.

Reply:
You can also just query this table, order by start ASC and iterate over the results and find any place where cur_result - last_result > 1 hour.

Reply:
Actually thinking here, we could just query `SELECT max(start) from StatisticRuns` and fill in tasks for that. We don't need to compare all old ones ?

Reply:
Fixed in 865b04c